apps/shell/app.js

```diff
@@ -56,67 +57,44 @@ app.use(process.env.PASSENGER_BASE_URI || '/', router);
 var server = new http.createServer(app);
 var wss = new WebSocket.Server({ noServer: true });
 
+whitelist = [];
+if (process.env.SSHHOST_WHITELIST){
+  whitelist = process.env.SSHHOST_WHITELIST.split(':');
+}
 
+glob.sync('/etc/ood/config/clusters.d/*.y*ml').forEach(function (file){
+  var data = yaml.safeLoad(fs.readFileSync(file));
+  const {
+    v2: {
+      metadata: {
+        hidden = true
+      },
+      login
+    }
+  } = data
 
-whitelist = process.env.SSHHOST_WHITELIST;
-if (whitelist === undefined){
-  whitelist = [];
-} else {
-  whitelist = whitelist.split(':');
-}
+  if (login){ //check login separately, doesn't exist in some config files
+    const{
+      host = 'default',
+      default: isDefault = false
+    } = login
```

Comment:
This can be moved back into line 72, if `login` doesn't exist nothing will break because there are defaults assigned in the destructuring assignment. I recommend using `localhost` as the default for host if its not defined.

```js
const {
  v2: {
    metadata: {
      hidden = true
    },
    login: {
      host = 'localhost',
      default: isDefault = false
    }
  }
} = data
```

Reply:
<img width="568" alt="Screen Shot 2020-05-29 at 11 22 29 PM" src="https://user-images.githubusercontent.com/56143351/83318479-484f9c80-a203-11ea-9289-2a9d1761ce0f.png">
It throws this error for files that don't have the login information

Reply:
Also I remember Eric saying something about having rules to disallow ssh to localhost

Reply:
These are cluster configs. If login does not exist, just as if the cluster is specified as hidden, this cluster config should be ignored. Since the number of cluster configs is small, this code might be easier to read if we separated filtering from adding. So something like (semi-pseudocode):

```js
var whitelist = glob.sync('/etc/ood/config/clusters.d/*.y*ml')
  .map(yml => yaml.safeLoad(fs.readFileSync(yml)))
  .filter(config => config.login_host_set && ! config.hidden)
}).reduce(((hosts, config) => config.default ? hosts.unshift(config.login) : hosts.push(config.login)), [])
```

Even if this were spaced out over several more lines for readability, this would be a lot less code.

Reply:
It looks like this code is trying to both filter and add. In any case, a majority of the code is spent destructuring the json object to usable values. config.v2.login.host would break the code if it doesn't exist
In addition, that would reassign the previously made whitelist if set in the environment (which could be added together later)